db/db_impl_open.cc

```diff
@@ -300,6 +298,35 @@ Status DBImpl::Directories::SetDirectories(
     }
   }
   assert(data_dirs_.size() == data_paths.size());
+
+  cf_data_dirs_.clear();
+  for (auto& column_family : column_families) {
+    if (!column_family.options.cf_paths.empty()) {
+      s = AddCFDirectories(env, column_family.name,
+                           column_family.options.cf_paths);
+      if (!s.ok()) {
+        return s;
+      }
+    }
+  }
+  return Status::OK();
+}
+
+Status DBImpl::Directories::AddCFDirectories(
+    Env* env, const std::string& cf_name,
+    const std::vector<DbPath>& cf_data_paths) {
+  Status s;
+  cf_data_dirs_[cf_name].clear();
```

Comment:
I kind of feel this info should be on `ColumnFamilyData`. It'd be easy to access as wherever we call `GetDataDir`, ColumnFamilyData` is available. This string-based lookup isn't great because (1) we don't have good failure handling for column family with same name, just an assert, which we should improve; (2) it's done while holding db mutex, which can block writes, particularly if user has long CF names.